rpmlb/recipe.py

```diff
@@ -65,24 +65,78 @@ def each_normalized_package(self):
-        if not recipe['name']:
-            raise ValueError('name is invalid in the recipe.')
-
-        if 'requires' in recipe:
-            if not isinstance(recipe['requires'], list):
-                raise ValueError('requires should be a list')
-
         if 'packages' not in recipe:
-            raise ValueError('packages is required in the recipe.')
+            raise ValueError('packages is required.')
         if not recipe['packages']:
-            raise ValueError('packages is invalid in the recipe.')
+            raise ValueError('packages is empty.')
         if not isinstance(recipe['packages'], list):
             raise ValueError('packages should be a list.')
+
+        for recipe_key in recipe:
+            if recipe_key == 'name':
+                if not recipe[recipe_key]:
+                    raise ValueError('{0} is empty.'.
+                                     format(recipe_key))
+
+                if not isinstance(recipe[recipe_key], str):
+                    raise ValueError('{0} shoule be a string.'.
+                                     format(recipe_key))
+            elif recipe_key == 'requires':
+                if not recipe[recipe_key]:
+                    raise ValueError('{0} is empty.'.
+                                     format(recipe_key))
+                if not isinstance(recipe[recipe_key], list):
+                    raise ValueError('{0} should be a list.'.
+                                     format(recipe_key))
+            elif recipe_key == 'packages':
+                pass
+            else:
+                raise ValueError(
+                    'Unknown element {0} exists.'.
+                    format(recipe_key))
+
         for package_dict in self.each_normalized_package():
-            if not package_dict['name']:
-                raise ValueError('name is invalid in the package.')
+            name = package_dict['name']
+            if not name:
+                raise ValueError('name is empty in the package.')
+
+            for key in package_dict.keys():
+                if key == 'name':
+                    pass
+                elif key == 'macros' or key == 'replaced_macros':
+                    if not package_dict[key]:
+                        raise ValueError('{0} is empty in the pacakge: {1}.'.
+                                         format(key, name))
+
+                    if not isinstance(package_dict[key], dict):
+                        raise ValueError(
+                            '{0} should be a hash in the package: {1}.'.
+                            format(key, name))
+                elif key == 'cmd':
+                    if not package_dict[key]:
+                        raise ValueError('{0} is empty in the package: {1}.'.
+                                         format(key, name))
+                    if not (isinstance(package_dict[key], str) or
+                       isinstance(package_dict[key], list)):
```

Comment:
`isinstance` can take more tan one type to check. Equivalent code:

```py
if not isinstance(package_dict[key], (str, list)):
```

> If classinfo is a tuple of type objects (or recursively, other such tuples), return true if object is an instance of any of the types.

https://docs.python.org/3/library/functions.html#isinstance